logstash-core/lib/logstash/agent_plugin_manager.rb

```diff
@@ -0,0 +1,65 @@
+# encoding: utf-8
+require "logstash/agent"
+
+module LogStash module AgentPluginManager
+  DEFAULT_AGENT_NAME = :default
+  class DuplicatePluginNameError < Error; end
+
+  REGISTRY = {}
+
+  # Reset plugin registry to just the default plugin
+  def self.reset!
+    REGISTRY.clear
+    REGISTRY[DEFAULT_AGENT_NAME] = LogStash::Agent
+  end
+
+  reset!
```

Comment:
@purbon could you please explain what kind of complicated to find out errors can result from this? I am not sure I understand your concern here. 

This seems to me a fine singleton pattern implementation, `reset!` is only called at module load time. something I am missing?